File: solutions/cpp/luhn/1/luhn.cpp
#include "luhn.h"
#include <algorithm>
#include <map>
#include <cctype>
namespace luhn {

// TODO: add your solution here
    static const std::map<char, std::pair<int, int>> DIGIT_TO_VALUE{
        {'0', {0, 0}},
        {'1', {1, 2}},
        {'2', {2, 4}},
        {'3', {3, 6}},
        {'4', {4, 8}},
        {'5', {5, 1}},
        {'6', {6, 3}},
        {'7', {7, 5}},
        {'8', {8, 7}},
        {'9', {9, 9}},
    };

    bool valid(std::string numbers) {
        numbers.erase(std::remove_if(numbers.begin(), numbers.end(), std::isspace), numbers.end());

        if (numbers.length() <= 1)
            return false;

        int sum{ 0 };
        bool make_double{ false };
        for (auto crit = numbers.rbegin(); crit != numbers.rend(); ++crit) {
            if (!std::isdigit(*crit))
                return false;

            const auto& values = DIGIT_TO_VALUE.at(*crit);
            if (make_double) {
                sum += values.second;
            } else {
                sum += values.first;
            }
            make_double = !make_double;
        }
        return sum % 10 == 0;
    }

}  // namespace luhn

File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num.replace(" ", "")

    @staticmethod
    def double_less_9(ch: str) -> int:
        value = int(ch) * 2
        return value if value <= 9 else value - 9

    def valid(self) -> bool:
        if len(self.card_num) <= 1:
            return False
        if not self.card_num.isdigit():
            return False

        sum1 = sum([Luhn.double_less_9(x) for x in self.card_num[-2::-2]])
        sum2 = sum([int(x) for x in self.card_num[-1::-2]])
        return (sum1 + sum2) % 10 == 0
